### PR TITLE
yolov7 tutorial

### PR DESCRIPTION
**Pull request by eaidova**

### CODE REVIEW

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 1]
```diff
@@ -0,0 +1,706 @@
...
+    "    --topk-all 100 --iou-thres 0.65 --conf-thres 0.35 \\\n",
+    "    --img-size 640 640 --max-wh 640\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Convert ONNX Model to OpenVINO Intermideate Representation\n",
+    "While ONNX models are directly supported by OpenVINO™, it can be useful to convert them to IR format to take advantage of OpenVINO optimization tools and features.\n",
```
Comment (ryanloney) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:171]:
```suggestion
    "While ONNX models are directly supported by OpenVINO runtime, it can be useful to convert them to IR format to take advantage of OpenVINO optimization tools and features.\n",
```

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+    "`mo.convert` function can be used for converting model using OpenVINO Model Optimizer capabilities. \n",
```
Comment (ryanloney) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:172]:
```suggestion
    "The `mo.convert` Python function can be used to convert a model using OpenVINO Model Optimizer. \n",
```

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 3]
```diff
@@ -0,0 +1,706 @@
...
+    "It returns of instance OpenVINO Model class, which is ready to use in python interface and can be serialized to IR for future execution."
```
Comment (ryanloney) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:173]:
```suggestion
    "The function returns instance of OpenVINO Model class, which is ready to use in Python interface but can also be serialized to OpenVINO IR format for future execution."
```

Comment (ryanloney) [reply notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:173]: Hope this makes sense. Tried to make it read a bit easier

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 4]
```diff
@@ -0,0 +1,706 @@
...
+    "## Verify model accuracy"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Download dataset\n",
+    "\n",
+    "Yolo V7 tiny pretrained on COCO dataset, in order to evaluate model accuracy we need to download it. According to instruction, provided in model repo, we also need to download annotation in prepared by model author format for using original model evaluation scripts"
```
Comment (ryanloney) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:389]:
```suggestion
    "YOLOv7 tiny is pre-trained on the COCO dataset, so in order to evaluate the model accuracy we need to download it. According to the instructions provided in the YOLOv7 repo, we also need to download annotations in the format used by the model author for use with the original model evaluation scripts"
```

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 5]
```diff
@@ -0,0 +1,706 @@
...
+    "                                       prefix=colorstr(f'{task}: '))[0]\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Define validation function\n",
+    "\n",
+    "We will reuse validation metrics provided in model repo with adoption to our case (removing extra steps). Original model evaluation procedure can be found in this [file](https://github.com/WongKinYiu/yolov7/blob/main/test.py)"
```
Comment (ryanloney) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:456]:
```suggestion
    "We will reuse validation metrics provided in the YOLOv7 repo with a modification for our case (removing extra steps). The original model evaluation procedure can be found in this [file](https://github.com/WongKinYiu/yolov7/blob/main/test.py)"
```

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 6]
```diff
@@ -0,0 +1,706 @@
...
+   "source": [
+    "int8_result = test(data=data, model=compiled_model, dataloader=dataloader)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Compare Performance of the Original and Quantized Models\n",
+    "Finally, use [Benchmark Tool](https://docs.openvino.ai/latest/openvino_inference_engine_tools_benchmark_tool_README.html) to measure the inference performance of the `FP16` and `INT8` models.\n",
```
Comment (ryanloney) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:654]:
```suggestion
    "Finally, use the OpenVINO [Benchmark Tool](https://docs.openvino.ai/latest/openvino_inference_engine_tools_benchmark_tool_README.html) to measure the inference performance of both the `FP16` and `INT8` models.\n",
```

**Review by ryanloney — CHANGES_REQUESTED** [review]
@eaidova I made a few suggestions. Not all cells run for me on Windows. I assume this is still WIP?

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 7]
```diff
@@ -0,0 +1,706 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Convert Yolo V7 to ONNX and OpenVINO™ IR\n",
+    "\n",
+    "This short tutorial demonstrates step-by-step instruction how to convert Pytorch Yolo V7 to OpenVINO IR. \n",
```
Comment (AlexKoff88) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:9]: It's not a "short tutorial", for sure)
I would rephrase it to:
```
This tutorial demonstrates step-by-step instructions on how to run and optimize *PyTorch Yolo V7 with OpenVINO.

It consists of the following steps:
```

**Comment by eaidova** [comment]
> @eaidova I made a few suggestions. Not all cells run for me on Windows. I assume this is still WIP?

I updated notebook code, now it should work

File: notebooks/225-yolov7-optimization/README.md [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# PyTorch YoloV7 model to OpenVINO™ Model Conversion and Optimization Tutorial
```
[the remark this answers is not in the document]
Comment (eaidova) [reply notebooks/225-yolov7-optimization/README.md]: updated

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb (hunk 7 above)
Comment (eaidova) [reply notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb]: Updated

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 9]
```diff
@@ -0,0 +1,706 @@
...
+    "- Compare performance of the FP32 and quantized models."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Get Pytorch model\n",
+    "\n",
+    "Generally, PyTorch model represents instance of torch.nn.Module class, iniatilized by state dictionary with model weights.\n",
```
[the remark this answers is not in the document]
Comment (eaidova) [reply notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb]: agree, added

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 10]
```diff
@@ -0,0 +1,706 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The most important parameters:\n",
+    "* `--weights` - path to model weigths checkpoint\n",
+    "* `--img-size` - size of input image for onnx tracing\n",
+    "\n",
+    "As ONNX is less flexible format then PyTorch, there is also opportunity to setup configurable parameters for results postprocessing included in model:\n",
```
[the remark this answers is not in the document]
Comment (eaidova) [reply notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb]: probably, it makes sense to describe at least provide info that difference in output format and hardcoded parameters of postprocessing, which can not be reconfigured any more (for accuracy measurement and preduction used different thresholds for confidence and overlap). The reason why this info can  be useful here, there is official tutorial which export onnx model within (it is not very informative :(), so user should understand why we choose own way and what should be done to enable model with postprocessing

**Review by AlexKoff88 — APPROVED** [review]

**Comment by eaidova** [comment]
@ryanloney could you please take a look one more time?

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 11]
```diff
@@ -0,0 +1,820 @@
...
+    "### Preprocessing\n",
+    "\n",
+    "Model input is tensor with shape `[1, 3, 640, 640]` in `N, C, H, W` format, where\n",
+    "* `N` - number of images in batch (batch size)\n",
+    "* `C` - image channels\n",
+    "* `H` - image height\n",
+    "* `W` - image width\n",
+    "\n",
+    "Model expects images in RGB channels format and normalized in [0, 1] range.\n",
+    "For resizing images to fit model size `letterbox` resize approach is used where the aspect ratio of width and height is preserved. It defined in yolov7 repository.\n",
```
Comment (zhuo-yoyowz) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:227]: It is defined in yolov7 repository.\n",

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 12]
```diff
@@ -0,0 +1,820 @@
...
+    "import numpy as np\n",
+    "import torch\n",
+    "import cv2\n",
+    "from utils.datasets import letterbox\n",
+    "from utils.plots import plot_one_box\n",
+    "\n",
+    "\n",
+    "def preprocess_image(img0: np.ndarray):\n",
+    "    \"\"\"\n",
+    "    Preprocess image according YOLOv7 input requirements. \n",
```
Comment (zhuo-yoyowz) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:246]: Preprocess image according to YOLOv7 input requirements.

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 13]
```diff
@@ -0,0 +1,820 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Optimize model using NNCF Postrainging Quantization API\n",
+    "\n",
+    "[NNCF](https://github.com/openvinotoolkit/nncf) provides a suite of advanced algorithms for Neural Networks inference optimization in OpenVINO with minimal accuracy drop.\n",
+    "We will use 8-bit quantization in post-training mode (without the fine-tuning pipeline) to optimize YOLOv7.\n",
+    "\n",
+    "> **Note**: NNCF Post-training Quantization is available as a preview feature in OpenVINO 2022.3 release.\n",
```
Comment (zhuo-yoyowz) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:643]: Does it mean this part could work after the release of 2022.3? Could it work with 22.2 as well? Thanks~

**Review by zhuo-yoyowz — APPROVED** [review]

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb (hunk 13 above)
Comment (eaidova) [reply notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:643]: it should works with pot in 2022.2 openvino-dev and NNCF from github if If my understanding correct (I tested it sometime ago, but it was on early POC stage).
@AlexKoff88  is it still true?

Just for clarification, this PR is targeted to `develop` branch for upcoming release, that is why readme provides info about 2022.3

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb (hunk 12 above)
Comment (eaidova) [reply notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb]: fixed, thanks

**Comment by eaidova** [comment]
@ryanloney can these changes be merged?

**Review by ryanloney — APPROVED** [review]

File: notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb [hunk 14]
```diff
@@ -0,0 +1,821 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from openvino.tools import mo\n",
+    "from openvino.runtime import serialize\n",
+    "\n",
+    "model = mo.convert(input_model='model/yolov7-tiny.onnx')\n",
```
Comment (slyalin) [inline notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:200]: In the recent master, we renamed it to `convert_model`, and you don't need to specify `input_model` anymore (but you still can if you want).

Comment (eaidova) [reply notebooks/225-yolov7-optimization/225-yolov7-optimiization.ipynb:200]: I still waiting package which will include this changes, after that I update...
